client-device-connection-infinispan/src/main/java/org/eclipse/hono/deviceconnection/infinispan/client/CacheBasedDeviceConnectionInfo.java

```diff
@@ -568,29 +590,60 @@ private String checkAdapterInstanceId(
         if (adapterInstanceId != null) {
             final AdapterInstanceStatus status = adapterInstanceStatusProvider.getStatus(adapterInstanceId);
             if (status == AdapterInstanceStatus.DEAD) {
-                LOG.debug("ignoring found adapter instance id, belongs to already terminated container [tenant: {}, device-id: {}, adapter-instance-id: {}]",
+                LOG.debug(
+                        "ignoring found adapter instance id, belongs to already terminated container [tenant: {}, device-id: {}, adapter-instance-id: {}]",
                         tenantId, deviceId, adapterInstanceId);
-                span.log("ignoring found adapter instance id [" + adapterInstanceId + "], belongs to already terminated container");
-                cache.remove(getAdapterInstanceEntryKey(tenantId, deviceId), adapterInstanceId)
+                span.log("ignoring found adapter instance id [" + adapterInstanceId
+                        + "], belongs to already terminated container");
+                final Future<Void> listenerResult;
+                if (deviceToAdapterMappingErrorListener != null) {
+                    listenerResult = deviceToAdapterMappingErrorListener.onObsoleteEntryFound(tenantId, deviceId,
+                            adapterInstanceId,
+                            span);
+                } else {
+                    listenerResult = Future.succeededFuture();
+                }
+                return listenerResult
                         .onSuccess(removed -> {
```

Comment:
Variable name `removed` is misleading here.

Reply:
agree, will change that in the next round